### PR TITLE
[CSI 0.3.0 3/4] csi-cephfs 0.3.0

### PR DESCRIPTION
**Pull request by gman0**
This is the third PR for the ceph-csi upgrade to 0.3.0, part of #47

What has changed:
* implemented `NodeStageVolume`, `NodeUnstageVolume`
* fixes https://github.com/ceph/ceph-csi/issues/48
* got rid of `volumeidentifier.go`, csi-cephfs no longer relies on volume ID semantics
* cleaned up keyring generator
* other refactoring, cleaning up

### CODE REVIEW

**Comment by rootfs** [comment]
@gman0 a general question. I am not convinced a local controller cache is robust. If the node is down and provisioner is relocated, all conf  on the old node are then gone.

A probably solution i am speculating is to persist these conf in api server, if the provisioner is deployed on k8s.

File: pkg/cephfs/cephconf.go [hunk 1]
```diff
@@ -35,16 +35,6 @@ fuse_set_user_groups = false
 
 const cephKeyring = `[client.{{.UserId}}]
 key = {{.Key}}
-caps mds = "allow rw path={{.RootPath}}"
```
[the remark this answers is not in the document]
Comment (gman0) [reply pkg/cephfs/cephconf.go]: AFAIK everything except the `key` is ignored anyway - regardless of what the other fields contain, so I've removed this dead code

File: pkg/cephfs/controllercache.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+type controllerCacheMap map[volumeID]*controllerCacheEntry
+
+var (
+	ctrCache    = make(controllerCacheMap)
+	ctrCacheMtx sync.Mutex
+)
+
+// Load all .json files from controllerCacheRoot into ctrCache
+// Called from driver.go's Run()
+func loadControllerCache() error {
```
[the remark this answers is not in the document]
Comment (gman0) [reply pkg/cephfs/controllercache.go]: only once, when the driver is instantiated

**Comment by rootfs** [comment]
@gman0 thanks!

**Comment by gman0** [comment]
I agree this needs to get better - as it is now it's basically useless, as you've said. I'm not sure how is the passing of CO-specific information doing, people are having mixed feelings about it. At least in this case, all that is needed is to `DeleteVolumeRequest` pass the `attributes` field for the volume, again I'm not sure why this is not included in the CSI spec.

**Pull request merged**